src/wmtk/PolygonMesh.hpp

```diff
@@ -56,59 +28,165 @@ class PolygonMesh : public Mesh
+     *
+     * @param tuple: tuple to check
+     * @param type: primitive of the tuple to check
+     * @return true iff the connectivity corresponding to the primitive is valid
+     */
+    bool is_local_connectivity_valid(const Tuple& tuple, PrimitiveType type) const;
 
     /**
-     * @brief Check if the face is a hole (i.e., not part of the surface)
+     * @brief Check if a face is a hole (i.e., not part of the surface)
+     *
+     * @param f_tuple: tuple specifying a face
+     * @return true iff the face is a hole
      */
     bool is_hole_face(const Tuple& f_tuple) const;
 
-    bool is_boundary(const Tuple& tuple) const override;
-    bool is_boundary_vertex(const Tuple& tuple) const override;
-    bool is_boundary_edge(const Tuple& tuple) const override;
-
-
-    bool is_valid(const Tuple& tuple, ConstAccessor<long>& hash_accessor) const override;
+    /**
+     * @brief Initialize a polygon mesh from a full halfedge representation with implicit opp.
+     *
+     * @param next: size #he vector, next halfedge id
+     * @param prev: size #he vector, prev halfedge id
+     * @param to: size #he vector, halfedge vertex tip id
+     * @param out: size #v vector, arbitrary halfedge id outgoing from vertex
+     * @param he2f: size #he vector, face id adjacent to halfedge
+     * @param f2he: size #f vector, arbitrary halfedge id adjacent to face
+     */
+    void initialize(
+        Eigen::Ref<const VectorXl> next,
+        Eigen::Ref<const VectorXl> prev,
+        Eigen::Ref<const VectorXl> to,
+        Eigen::Ref<const VectorXl> out,
+        Eigen::Ref<const VectorXl> he2f,
+        Eigen::Ref<const VectorXl> f2he);
 
+    /**
+     * @brief Initialize a polygon mesh from a minimal halfedge representation with implicit opp.
+     *
+     * Face and vertex ids are generated implicitly as needed.
+     *
+     * @param next: size #he vector, next halfedge id
+     */
     void initialize(Eigen::Ref<const VectorXl> next);
+
+    /**
+     * @brief Initialize a polygon mesh from an FV representation.
+     *
+     * @param F: list of lists of face vertices
+     */
     void initialize_fv(const std::vector<std::vector<long>>& F);
+
+    /**
+     * @brief Initialize a triangle mesh from an FV representation.
+     *
+     * @param F: $fx3 matrix of face vertices
+     */
     void initialize_fv(Eigen::Ref<const RowVectors3l> F);
 
 protected:
     long id(const Tuple& tuple, PrimitiveType type) const override;
 
-
 protected:
-    attribute::MeshAttributeHandle<long> m_en_handle;
-    attribute::MeshAttributeHandle<long> m_ep_handle;
+    attribute::MeshAttributeHandle<long> m_next_handle; // HalfEdge -> next HalfEdge
+    attribute::MeshAttributeHandle<long> m_prev_handle; // HalfEdge -> previous HalfEdge
+
+    attribute::MeshAttributeHandle<long> m_to_handle; // HalfEdge -> Vertex at tip
+    attribute::MeshAttributeHandle<long> m_out_handle; // Vertex -> any outgoing HalfEdge
```

Comment:
why is this not named something like vh? you can alias it as well
same with the above to. just to keep names similar. If you want to create an alias you can use member functions to induce your expected  names `MeshAttributeHandle to_handle() const { return m_hv_handle; }`

Reply:
I considered using `hv`, but the faces and vertices aren't really symmetric in the halfedge representation, and there's an ambiguity with `hv` that isn't present with `hf`. Since a halfedge is (more or less) a face-edge pair, each halfedge has an unambiguous associated face. However, there are two consistent choices for the vertex that can be associated to the halfedge by `hv`: one at the tip of the edge and one at the tail. Using `to` and `out` removes the ambiguity.

The interface, on the other hand, functions like a quad edge representation where faces and vertices are actually symmetric. The tuple -> face ID map directly uses the `hf` attribute for the global id of the tuple halfedge, whereas the tuple -> vertex ID map has more complex logic that depends on the vid (i.e., orientation) data of the tuple. I think using `to` instead of `hv` makes it a bit more clear that there is this asymmetry in the underlying halfedge representation, and the vertex ID stored in this attribute depends on a default orientation of the halfedge mesh that must be accounted for in the quad edge tuple representation.

That being said, if you still want to use `vh` and `hv` for consistency with the `ff`, `fv`, etc. maps in `TriMesh`, I can make the suggested change.

Reply:
(Discussed in person)